Android/Samples/libraries/kml/src/main/java/com/microsoft/maps/kml/KMLParser.java

```diff
@@ -59,7 +51,79 @@ public static MapElementLayer parse(@NonNull String kml) throws KMLParseExceptio
     }
   }
 
-  MapElementLayer internalParse(@NonNull String kml, @NonNull MapFactories factory) {
+  @VisibleForTesting
+  @NonNull
+  MapElementLayer internalParse(@NonNull String kml, @NonNull MapFactories factory)
+      throws XmlPullParserException, IOException, KMLParseException {
+    mFactory = factory;
+    mLayer = mFactory.createMapElementLayer();
+    try (InputStream stream = new ByteArrayInputStream(kml.getBytes(UTF_8))) {
+      mParser.setInput(stream, null);
+      mParser.nextTag();
+      mNameSpace = mParser.getNamespace();
+      parsePlacemark();
+    }
+    return mLayer;
+  }
+
+  private void parsePlacemark() throws IOException, XmlPullParserException, KMLParseException {
+    while (mParser.next() != XmlPullParser.END_TAG) {
+      if (mParser.getEventType() != XmlPullParser.START_TAG) {
+        continue;
+      }
+      String type = mParser.getName();
+      if (type.equals("Placemark")) {
+        // used for MapIcon Title
+        String name = parseName();
+      }
+    }
+  }
+
+  @Nullable
+  private String parseName() throws IOException, XmlPullParserException, KMLParseException {
+    while (mParser.next() != XmlPullParser.END_TAG) {
+      if (mParser.getEventType() != XmlPullParser.START_TAG) {
+        continue;
+      }
+      String type = mParser.getName();
+      if (type.equals("name")) {
+        mParser.require(XmlPullParser.START_TAG, mNameSpace, "name");
+        String title = parseText();
+        mParser.require(XmlPullParser.END_TAG, mNameSpace, "name");
+        return title;
+      } else {
+        skip();
+      }
+    }
     return null;
   }
+
+  @Nullable
+  private String parseText() throws IOException, XmlPullParserException {
+    String result = null;
+    if (mParser.next() == XmlPullParser.TEXT) {
+      result = mParser.getText();
+      mParser.nextTag();
+    }
+    return result;
+  }
+
+  private void skip() throws XmlPullParserException, IOException, KMLParseException {
```

Comment:
This function looks like it's only valid if you're at a start tag; document this requirement please.